src/Microsoft.AspNet.HttpOverrides/OverrideHeaderMiddleware.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+
+namespace Microsoft.AspNet.HttpOverrides
+{
+    public class OverrideHeaderMiddleware
+    {
+        private const string XForwardedFor = "X-Forwarded-For";
+        private const string XForwardedHost = "X-Forwarded-Host";
+        private const string XForwardedProto = "X-Forwarded-Proto";
+        private readonly OverrideHeaderMiddlewareOptions _options;
+        private readonly RequestDelegate _next;
+
+        public OverrideHeaderMiddleware(RequestDelegate next, OverrideHeaderMiddlewareOptions options)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (next == null)
+            {
+                throw new ArgumentNullException(nameof(_next));
+            }
+            _options = options;
+            _next = next;
+        }
+
+        public Task Invoke(HttpContext context)
+        {
+            if ((_options.ForwardedOptions & ForwardedHeaders.XForwardedFor) != 0)
```

Comment:
A couple of things here worry me - 1) We're throwing away the original RemoteIpAddress, which we should just map into another header and 2) are we really considering scenarios where there are multiple different forwarding headers? Because we're making up a precedence order if this happens. You know people will just use All, because they're lazy.